agents/aggregator/aggregator_agent.py

```diff
@@ -18,463 +13,511 @@ class Provider:
+                                      options={'match': 'wildcard'})
 
-        new_file_time = True
+        record_on_start = (args.initial_state == 'record')
+        self.agent.register_process('record', self.start_aggregate, self.stop_aggregate,
+                                    startup=record_on_start)
 
-        self.aggregate = True
+    def data_handler(self, _data):
+        data, feed = _data
 
-        while self.aggregate:
-            time.sleep(1)
-
-            if new_file_time:
-                if self.writer is not None:
-                    self.end_file()
-                self.start_file(data_dir)
-                file_start_time = time.time()
-
-
-            # Removes old providers if new ones exist
-            keys = sorted(self.prov_ids, key=lambda x: x[1], reverse=True)
-            feeds = []
-            for k in keys:
-                if k[0] in feeds:
-                    prov = self.providers[self.prov_ids[k]]
-                    if not prov.remove:
-                        print("Newer {} feed found".format(k[0]))
-                        prov.remove = True
-                else:
-                    feeds.append(k[0])
-
-            to_remove = [p for _, p in self.providers.items() if p.remove]
-            if to_remove != []:
-                self.should_write_status = True
-
-            # If any providers have been removed, write those to disk first
-            # And remove provider from hksess
-            prov : Provider
-            for prov in to_remove:
-                if prov.frame_start_time is not None:
-                    with prov.lock:
-                        frame = prov.to_frame(self.hksess)
-                        prov.clear()
-                    self.writer(frame)
-
-                self.log.info("Removing provider {} with session_id {}"
-                              .format(prov.address, prov.session_id))
-                pid = prov.prov_id
-                self.hksess.remove_provider(pid)
-                del self.providers[pid]
-                del self.prov_ids[(prov.address, prov.session_id)]
-
-            # Then write status if we need to
-            if self.should_write_status:
-                self.write_status()
-
-            # Write to disk any active providers that have surpassed frame_length
-            for pid, prov in self.providers.items():
-                if prov.frame_start_time is None:
-                    continue
+        if not feed['record']:
+            return
 
-                if prov.remove or (time.time() - prov.frame_start_time > prov.frame_length):
-                    with prov.lock:
-                        frame = prov.to_frame(self.hksess)
-                        prov.clear()
-                    self.writer(frame)
+        address = feed['address']
+        sessid = feed['session_id']
+        frame_length = feed['agg_params']['frame_length']
 
-            self.writer.Flush()
+        pid = self.aggregator.pids.get((address, sessid))
```

Comment:
This is an unsafe access to self.aggregator internal variables -- you should use a method call, because that can be inspected for thread-safety in the internal context of the Aggregator class.

It's likely that you'll need to protect Aggregator.pids / aggregator.providers with some kind of lock.